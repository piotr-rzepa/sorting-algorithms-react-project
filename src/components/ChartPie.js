/*
 * Komponent bezklasowy będacy wykresem kołowym z podziałem algorytmów \
 * na kategorie względem rodzaju sortowania jaki stosują
 */

/* eslint-disable react/prop-types */
import React from 'react';
import { useSelector } from 'react-redux';
import { PieChart, Pie, Cell, Legend, Tooltip, LabelList } from 'recharts';
import selectMethodsOfSorting from '../selectors/sortingMethodSelector';

export default function CardChart(props) {
	const barColors = ['#33D69F', '#6F52ED', '#FF4C61', '#FFB800'];
	const algorithms = useSelector((state) => selectMethodsOfSorting(state));

	return (
		<div className="card-chart card-chart-small">
			<p> Algorithms by method of sorting</p>
			<hr />
			<PieChart width={props.width} height={props.height}>
				<Pie
					data={algorithms}
					dataKey="value"
					cx="50%"
					cy="50%"
					innerRadius={80}
					outerRadius={100}
					label
				>
					{algorithms.map((_, index) => (
						<Cell key={`cell-${index}`} fill={barColors[index]} />
					))}
					<LabelList dataKey="method" position="insideEnd" />
				</Pie>
				<Legend
					layout="horizontal"
					align="center"
					iconType="plainline"
					iconSize={60}
				/>
				<Tooltip
					formatter={(value, name) => (
						<p>
							Method: {algorithms[name].method}
							<br />
							Algortihms: {value}
						</p>
					)}
				/>
			</PieChart>
		</div>
	);
}
