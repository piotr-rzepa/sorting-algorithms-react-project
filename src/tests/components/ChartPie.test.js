import * as React from 'react';
import { shallow } from 'enzyme';
import * as reactRedux from 'react-redux';
import ChartPie from '../../components/ChartPie';
import { Tooltip } from 'recharts';

const data = [
	{
		id: 0,
		name: 'Bubble Sort',
		description:
			'Bubble sort, sometimes referred to as sinking sort, is a simple sorting algorithm',
		method: 'Exchanging',
		sortingTime: { best: 0, worst: 0, average: 0 }
	},
	{
		id: 1,
		name: 'Quick Sort',
		description: 'Quicksort is a divide-and-conquer algorithm.',
		method: 'Partitioning',
		sortingTime: { best: 0, worst: 0, average: 0 }
	}
];

describe('<ChartPie />', () => {
	const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
	const useContextMock = jest.spyOn(React, 'useContext');
	beforeEach(() => {
		useSelectorMock.mockClear();
		useContextMock.mockClear();
	});

	it('renders correctly', () => {
		const dummyDispatch = jest.fn();
		useContextMock.mockReturnValue(dummyDispatch);
		useSelectorMock.mockReturnValue(data);
		const wrapper = shallow(<ChartPie />);
		expect(wrapper).toMatchSnapshot();
	});

	it('Should format ToolTip correctly', () => {
		const dummyDispatch = jest.fn();
		useContextMock.mockReturnValue(dummyDispatch);
		useSelectorMock.mockReturnValue([{ name: 'Bubble Sort', method: 'XD' }]);
		const wrapper = shallow(<ChartPie />);
		const result = wrapper.find(Tooltip).prop('formatter')(10, 0);
		expect(result).toMatchSnapshot();
	});

	it('should choose correct black theme', () => {
		useContextMock.mockReturnValue(true);
		useSelectorMock.mockReturnValue(data);
		const wrapper = shallow(<ChartPie />);
		expect(wrapper.find('div.card-chart').get(0).props.style).toHaveProperty(
			'color',
			'white'
		);
	});

	it('should choose correct white theme', () => {
		useContextMock.mockReturnValue(false);
		useSelectorMock.mockReturnValue(data);
		const wrapper = shallow(<ChartPie />);
		expect(wrapper.find('div.card-chart').get(0).props.style).toHaveProperty(
			'backgroundColor',
			'white'
		);
	});
});
