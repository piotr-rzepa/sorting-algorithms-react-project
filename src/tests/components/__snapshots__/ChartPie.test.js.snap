// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChartPie /> Should format ToolTip correctly 1`] = `
<p>
  Method: 
  XD
  <br />
  Algortihms: 
  10
</p>
`;

exports[`<ChartPie /> renders correctly 1`] = `
<div
  className="card-chart card-chart-small"
  style={
    Object {
      "backgroundColor": "#606060",
      "boxShadow": "2px 2px 2px black",
      "color": "white",
      "height": undefined,
    }
  }
>
  <p>
     Algorithms by method of sorting
  </p>
  <hr />
  <ResponsiveContainer
    debounce={0}
    height="100%"
    width="100%"
  >
    <PieChart
      barCategoryGap="10%"
      barGap={4}
      cx="50%"
      cy="50%"
      endAngle={360}
      innerRadius={0}
      layout="centric"
      margin={
        Object {
          "bottom": 5,
          "left": 5,
          "right": 5,
          "top": 5,
        }
      }
      outerRadius="80%"
      reverseStackOrder={false}
      stackOffset="none"
      startAngle={0}
      syncMethod="index"
    >
      <Pie
        animationBegin={400}
        animationDuration={1500}
        animationEasing="ease"
        blendStroke={false}
        cx="50%"
        cy="50%"
        data={
          Array [
            Object {
              "description": "Bubble sort, sometimes referred to as sinking sort, is a simple sorting algorithm",
              "id": 0,
              "method": "Exchanging",
              "name": "Bubble Sort",
              "sortingTime": Object {
                "average": 0,
                "best": 0,
                "worst": 0,
              },
            },
            Object {
              "description": "Quicksort is a divide-and-conquer algorithm.",
              "id": 1,
              "method": "Partitioning",
              "name": "Quick Sort",
              "sortingTime": Object {
                "average": 0,
                "best": 0,
                "worst": 0,
              },
            },
          ]
        }
        dataKey="value"
        endAngle={360}
        fill="#808080"
        hide={false}
        innerRadius={80}
        isAnimationActive={true}
        label={true}
        labelLine={true}
        legendType="rect"
        minAngle={0}
        nameKey="name"
        outerRadius={100}
        paddingAngle={0}
        startAngle={0}
        stroke="#fff"
      >
        <Cell
          fill="#33D69F"
          key="cell-0"
        />
        <Cell
          fill="#6F52ED"
          key="cell-1"
        />
        <LabelList
          dataKey="method"
          position="insideEnd"
          valueAccessor={[Function]}
        />
      </Pie>
      <Legend
        align="center"
        iconSize={60}
        iconType="plainline"
        layout="horizontal"
        verticalAlign="bottom"
      />
      <Tooltip
        active={false}
        allowEscapeViewBox={
          Object {
            "x": false,
            "y": false,
          }
        }
        animationDuration={400}
        animationEasing="ease"
        contentStyle={Object {}}
        coordinate={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        cursor={true}
        cursorStyle={Object {}}
        filterNull={true}
        formatter={[Function]}
        isAnimationActive={true}
        itemStyle={Object {}}
        labelStyle={Object {}}
        offset={10}
        separator=" : "
        trigger="hover"
        useTranslate3d={false}
        viewBox={
          Object {
            "x1": 0,
            "x2": 0,
            "y1": 0,
            "y2": 0,
          }
        }
        wrapperStyle={Object {}}
      />
    </PieChart>
  </ResponsiveContainer>
</div>
`;
